src/train.py

```diff
@@ -0,0 +1,451 @@
+
+def pad_collate(batch):
+    """
+    Custom collate function to pad the sequences in the batch for DataLoader.
+
+    Args:
+        batch (list): A list of tuples containing the features and labels for each sample.
+
+    Returns:
+        A tuple containing the padded features and labels tensors.
+    """
+    max_seq_len = max([len(item[0]) for item in batch])
+
+    padded_features_batch = []
+    padded_labels_batch = []
+
+    for features, labels in batch:
+        seq_len = len(features)
+        pad_len = max_seq_len - seq_len
+        padded_features = torch.cat([features, torch.tensor([[-1e6] * features.shape[1]] * pad_len, dtype=torch.float)])
+        padded_labels = torch.cat([labels, torch.tensor([[-1] * labels.shape[1]] * pad_len, dtype=torch.float)])
+        padded_features_batch.append(padded_features)
+        padded_labels_batch.append(padded_labels)
+
+    padded_features_batch = torch.stack(padded_features_batch)
+    padded_labels_batch = torch.stack(padded_labels_batch)
+
+    return padded_features_batch, padded_labels_batch
+
+
+def masked_binary_cross_entropy_with_logits(logits, targets, mask, neg_to_pos_ratios):
+    """
+    Custom loss function that calculates the masked binary cross-entropy loss for the DL model.
+
+    Args:
+        logits (torch.Tensor): The logits output from the model.
+        targets (torch.Tensor): The ground truth labels.
+        mask (torch.Tensor): The mask for valid target values.
+        neg_to_pos_ratios (list): The list of negative to positive ratios for each task.
+
+    Returns:
+        A scalar value representing the masked binary cross-entropy loss.
+    """
+    losses = 0
+    for i in range(len(neg_to_pos_ratios)):
+        # Create a mask for valid target values (not equal to -1)
+        valid_targets_mask = targets[:, :, i] != -1
+        # Filter out invalid elements from logits and targets using the mask
+        filtered_logits = logits[:, :, i][valid_targets_mask]
+        filtered_targets = targets[:, :, i][valid_targets_mask]
+        # Calculate the loss
+        loss = nn.functional.binary_cross_entropy_with_logits(
+            filtered_logits, filtered_targets, reduction='none', pos_weight=torch.tensor(neg_to_pos_ratios[i])
+        )
+        masked_loss = loss * mask[:, :, i][valid_targets_mask]
+        losses += masked_loss.sum() / mask[:, :, i][valid_targets_mask].sum()
+
+    losses = losses / len(neg_to_pos_ratios)
+    return losses
+
+def calibrated_model(model, dataloader, neg_to_pos_ratios,output_size, device):
+    model.eval()
+    running_loss = [0.0] * output_size
+    all_labels = [[] for _ in range(output_size)]
+    all_preds = [[] for _ in range(output_size)]
+
+    with torch.no_grad():
+        for i, batch in enumerate(dataloader):
+            inputs, labels = batch
+            inputs, labels = (inputs.to(device, dtype=torch.float), labels.to(device, dtype=torch.float))
```

Comment:
Also, try to minimize repeated code. validate_dl has the same block of code as here. You can place that block of code into a predict function.